File: 6.ZigZagConversion.java
class Solution {
  public String convert(String s, int numRows) {
    int length = s.length();
    if (numRows > length || numRows <= 1)
      return s;

    int count = 0;
    int interval = 2 * numRows - 2;
    char[] zigZagChars = new char[length];

    for (int i = 0; i < numRows; i++) {
      int step = interval - 2 * i;
      for (int j = i; j < length; j += interval) {
        zigZagChars[count++] = s.charAt(j);
        if (step > 0 && step < interval && j + step < length) {
          zigZagChars[count++] = s.charAt(j + step);
        }
      }
    }

    return new String(zigZagChars);
  }
}

File: 6.ZigZagConversion.js
var convert = function (s, numRows) {
  len = s.length;
  if (numRows > len || numRows <= 1) {
    return s;
  }

  let count = 0,
    interval = 2 * numRows - 2,
    zigZagChars = [];

  for (let i = 0; i < numRows; i++) {
    let step = interval - 2 * i;
    for (let j = i; j < len; j += interval) {
      zigZagChars[count++] = s[j];
      if (step > 0 && step < interval && step < len) {
        zigZagChars[count++] = s[j + step];
      }
    }
  }

  return zigZagChars.join("");
};